File: sorting/quick-sort/quickSort.js
const split = (data, left, right) => {
    const pivot = data[right];

    let middlePoint = left - 1
    for (let i = left; i < right; i++) {
        if (data[i] < pivot) {
            middlePoint++
            [data[middlePoint], data[i]] = [data[i], data[middlePoint]]
        }
    }
    [data[middlePoint+1], data[right]] = [data[right], data[middlePoint+1]]

    return middlePoint + 1
};

const quickSoftAsc = (data, left, right) => {
    if (left < right) {
        const middlePoint = split(data, left, right)
        quickSoftAsc(data, left, middlePoint - 1)
        // middle point is in the final proper position
        quickSoftAsc(data, middlePoint + 1, right)
    }
};

const quickSort = (d) => quickSoftAsc(d, 0, d.length - 1);

const d = [50, 5, 7, 98, 4, 2, 5, 7, 9];
quickSort(d);
console.log(d);

File: sorting/quick-sort/quickSort.go
package main

import (
	"fmt"
)

func split(data []int, splitElement int) ([]int, []int) {
	dataLength := len(data)

	middlePoint := -1
	i := 0
	for ; i < len(data)-1; i++ {
		if data[i] <= splitElement {
			middlePoint++
			data[i], data[middlePoint] = data[middlePoint], data[i]
		}
	}

	data[middlePoint+1], data[dataLength-1] = data[dataLength-1], data[middlePoint+1]

	return data[:middlePoint+1], data[middlePoint+1:]
}

func quickSortAsc(data []int) []int {
	if len(data) <= 1 {
		return data
	}

	splitElement := data[len(data)-1]

	leftData, rightData := split(data, splitElement)

	quickSortAsc(leftData)
	quickSortAsc(rightData)

	return data
}

func main() {
	d := []int{50, 5, 7, 98, 4, 2, 5, 7, 9}
	resultAsc := quickSortAsc(d)
	fmt.Println(resultAsc)
}
